src/System.Net.NameResolution/src/System/Net/DNS.cs

```diff
@@ -0,0 +1,383 @@
+
+            //
+            // If IPv6 is not enabled (maybe config switch) but we've been
+            // given an IPv6 address then we need to bail out now.
+            //
+            else
+            {
+                if (address.AddressFamily == AddressFamily.InterNetworkV6)
+                {
+                    //
+                    // Protocol not supported
+                    //
+                    throw new SocketException((int)SocketError.ProtocolNotSupported);
+                }
+                //
+                // Use gethostbyaddr() to try to resolve the IP address
+                //
+                // End IPv6 Changes
+                //
+                return NameResolutionPal.GetHostByAddr(address);
+            }
+        } // InternalGetHostByAddress
+
+        /*****************************************************************************
+         Function :    gethostname
+
+         Abstract:     Queries the hostname from DNS
+
+         Input Parameters:
+
+         Returns: String
+        ******************************************************************************/
+
+        /// <devdoc>
+        ///    <para>Gets the host name of the local machine.</para>
+        /// </devdoc>
+        public static string GetHostName()
+        {
+            GlobalLog.Print("Dns.GetHostName");
+            return NameResolutionPal.GetHostName();
+        }
+
+        private class ResolveAsyncResult : ContextAwareResult
+        {
+            // Forward lookup
+            internal ResolveAsyncResult(string hostName, object myObject, bool includeIPv6, object myState, AsyncCallback myCallBack) :
+                base(myObject, myState, myCallBack)
+            {
+                this.hostName = hostName;
+                this.includeIPv6 = includeIPv6;
+            }
+
+            // Reverse lookup
+            internal ResolveAsyncResult(IPAddress address, object myObject, bool includeIPv6, object myState, AsyncCallback myCallBack) :
+                base(myObject, myState, myCallBack)
+            {
+                this.includeIPv6 = includeIPv6;
+                this.address = address;
+            }
+
+            internal readonly string hostName;
+            internal bool includeIPv6;
+            internal IPAddress address;
+        }
+
+        private static void ResolveCallback(object context)
+        {
+            ResolveAsyncResult result = (ResolveAsyncResult)context;
+            IPHostEntry hostEntry;
+            try
+            {
+                if (result.address != null)
+                {
+                    hostEntry = InternalGetHostByAddress(result.address, result.includeIPv6);
+                }
+                else
+                {
+                    hostEntry = InternalGetHostByName(result.hostName, result.includeIPv6);
+                }
+            }
+            catch (Exception exception)
+            {
+                if (exception is OutOfMemoryException)
```

Comment:
Port artifact: this was performing more exception type checks initially. (Such as ThreadAbortException and StackOverflowException which have been now removed.) Switching to the proposed pattern.